File: app/src/main/java/com/example/rooster/veterinary/consultations/VetConsultationScreen.kt
package com.example.rooster.veterinary.consultations

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.PersonAdd
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.random.Random

// --- Data Classes ---
data class VetProfile(
    val vetId: String,
    val name: String,
    val specialty: String,
    var availability: Map<String, List<String>>, // Date string to list of time slots (e.g., "HH:mm")
    val consultationFee: Double
)

data class Appointment(
    val appointmentId: String,
    val vetId: String,
    val vetName: String,
    val userId: String,
    val patientName: String,
    val appointmentDateTime: Date,
    val reason: String,
    var status: String // booked, completed, cancelled, in_progress
) {
    fun getFormattedDateTime(): String = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault()).format(appointmentDateTime)
}

data class PatientQueueEntry(
    val queueId: String,
    val userId: String,
    val patientName: String,
    val arrivalTime: Date,
    val reason: String,
    var assignedVetId: String? = null,
    var assignedVetName: String? = null, // For display
    var status: String = "waiting" // waiting, assigned, in_consultation
) {
    fun getFormattedArrivalTime(): String = SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(arrivalTime)
}

// --- ViewModel ---
class VetConsultationViewModel : ViewModel() {
    private val _vets = MutableStateFlow<List<VetProfile>>(emptyList())
    val vets: StateFlow<List<VetProfile>> = _vets

    private val _appointments = MutableStateFlow<List<Appointment>>(emptyList())
    val appointments: StateFlow<List<Appointment>> = _appointments

    private val _patientQueue = MutableStateFlow<List<PatientQueueEntry>>(emptyList())
    val patientQueue: StateFlow<List<PatientQueueEntry>> = _patientQueue

    val specialties = listOf("General Practice", "Surgery", "Dermatology", "Livestock Specialist", "Cardiology")
    val mockPatientNames = listOf("Bessie", "Charlie", "Daisy", "Rocky", "Lucy", "Max", "Speckles")
    val mockReasons = listOf("General check-up", "Vaccination", "Skin issue", "Limping", "Not eating", "Urgent concern")


    init {
        loadMockData()
    }

    private fun generateMockAvailability(numDays: Int = 7): Map<String, List<String>> {
        val availability = mutableMapOf<String, List<String>>()
        val today = Calendar.getInstance()
        for (i in 0 until numDays) {
            val dayCal = today.clone() as Calendar
            dayCal.add(Calendar.DATE, i)
            if (dayCal.get(Calendar.DAY_OF_WEEK) == Calendar.SATURDAY || dayCal.get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY) continue

            val timeSlots = mutableListOf<String>()
            for (j in 0..4) { // 5 slots per day
                val hour = 9 + j
                if (Random.nextBoolean()) { // Slot available?
                    timeSlots.add(String.format("%02d:00", hour))
                }
            }
            if (timeSlots.isNotEmpty()) {
                availability[SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(dayCal.time)] = timeSlots
            }
        }
        return availability
    }

    private fun loadMockData() {
        val vetNames = listOf("Dr. Alice Smith", "Dr. Bob Johnson", "Dr. Carol White", "Dr. David Brown")
        val tempVets = vetNames.mapIndexed { i, name ->
            VetProfile(
                vetId = "vet${(i + 1).toString().padStart(3, '0')}",
                name = name,
                specialty = specialties.random(),
                availability = generateMockAvailability(),
                consultationFee = Random.nextDouble(50.0, 150.0)
            )
        }
        _vets.value = tempVets

        val tempAppointments = mutableListOf<Appointment>()
        // Create some mock appointments
        _appointments.value = tempAppointments.sortedBy { it.appointmentDateTime }

        val tempQueue = (1..3).map { i ->
            PatientQueueEntry(
                queueId = "q_${Date().time}_$i",
                userId = "farmer${Random.nextInt(1,5).toString().padStart(3,'0')}",
                patientName = mockPatientNames.random(),
                arrivalTime = Date(System.currentTimeMillis() - Random.nextLong(5, 30) * 60000L),
                reason = mockReasons.random()
            )
        }.toMutableList()
        _patientQueue.value = tempQueue.sortedBy { it.arrivalTime }
    }

    fun bookAppointment(vetId: String, userId: String, patientName: String, dateStr: String, timeStr: String, reason: String): Boolean {
        val vet = _vets.value.find { it.vetId == vetId } ?: return false
        val slotsForDate = vet.availability[dateStr] ?: return false
        if (!slotsForDate.contains(timeStr)) return false

        // Remove slot (simplified)
        val updatedAvailability = vet.availability.toMutableMap()
        updatedAvailability[dateStr] = slotsForDate.filterNot { it == timeStr }
        _vets.update { list -> list.map { if (it.vetId == vetId) it.copy(availability = updatedAvailability) else it } }

        val appointmentDateTime = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault()).parse("$dateStr $timeStr") ?: Date()
        val newAppointment = Appointment(
            appointmentId = "appt_${Date().time}", vetId = vetId, vetName = vet.name, userId = userId,
            patientName = patientName, appointmentDateTime = appointmentDateTime, reason = reason, status = "booked"
        )
        _appointments.update { (it + newAppointment).sortedBy { appt -> appt.appointmentDateTime } }
        return true
    }

    fun addToQueue(userId: String, patientName: String, reason: String) {
        val newEntry = PatientQueueEntry(
            queueId = "q_${Date().time}_${Random.nextInt(100)}", userId = userId, patientName = patientName,
            arrivalTime = Date(), reason = reason
        )
        _patientQueue.update { (it + newEntry).sortedBy { entry -> entry.arrivalTime } }
    }

    fun assignQueuedPatient(queueId: String, vetId: String) {
        val vet = _vets.value.find { it.vetId == vetId } ?: return
        _patientQueue.update { queue ->
            queue.map {
                if (it.queueId == queueId && it.assignedVetId == null)
                    it.copy(assignedVetId = vetId, assignedVetName = vet.name, status = "assigned")
                else it
            }
        }
    }
}

// --- Composable Screen ---
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun VetConsultationScreen(viewModel: VetConsultationViewModel = androidx.lifecycle.viewmodel.compose.viewModel()) {
    val vets by viewModel.vets.collectAsState()
    val appointments by viewModel.appointments.collectAsState()
    val patientQueue by viewModel.patientQueue.collectAsState()

    var showBookingDialog by remember { mutableStateOf(false) }
    var vetForBooking by remember { mutableStateOf<VetProfile?>(null) }
    var showAddToQueueDialog by remember { mutableStateOf(false) }

    if (showBookingDialog && vetForBooking != null) {
        BookingDialog(vet = vetForBooking!!, viewModel = viewModel, onDismiss = { showBookingDialog = false; vetForBooking = null })
    }
    if(showAddToQueueDialog) {
        AddToQueueDialog(viewModel = viewModel, onDismiss = { showAddToQueueDialog = false })
    }

    Scaffold(
        topBar = { TopAppBar(title = { Text("Vet Consultation") }, colors = TopAppBarDefaults.topAppBarColors(containerColor = Color(0xFF00ACC1))) }, // Cyan
        floatingActionButton = {
            FloatingActionButton(onClick = { showAddToQueueDialog = true }) {
                Icon(Icons.Filled.PersonAdd, "Add to Walk-in Queue")
            }
        }
    ) { paddingValues ->
        LazyColumn(modifier = Modifier.padding(paddingValues).padding(16.dp)) {
            item { SectionTitle("Available Veterinarians") }
            items(vets, key = {it.vetId}) { vet -> VetCard(vet) { selectedVet -> vetForBooking = selectedVet; showBookingDialog = true } }

            item { SectionTitle("Upcoming Appointments (${appointments.filter{it.status == "booked" && it.appointmentDateTime >= Date()}.size})") }
            val upcomingAppointments = appointments.filter{it.status == "booked" && it.appointmentDateTime >= Date()}.take(5)
            if(upcomingAppointments.isEmpty()) item { EmptyState("No upcoming appointments.")}
            items(upcomingAppointments, key = {it.appointmentId}) { appt -> AppointmentCard(appt) }

            item { SectionTitle("Patient Queue (${patientQueue.size})") }
             if(patientQueue.isEmpty()) item { EmptyState("Patient queue is empty.")}
            items(patientQueue, key = {it.queueId}) { entry -> PatientQueueCard(entry, vets, viewModel) }
        }
    }
}

@Composable
fun SectionTitle(title: String) = Text(title, style = MaterialTheme.typography.headlineSmall, modifier = Modifier.padding(top = 16.dp, bottom = 8.dp))
@Composable
fun EmptyState(message: String) = Text(message, modifier = Modifier.padding(vertical = 8.dp).fillMaxWidth(), textAlign = androidx.compose.ui.text.style.TextAlign.Center)


@Composable
fun VetCard(vet: VetProfile, onBookClick: (VetProfile) -> Unit) {
    Card(Modifier.fillMaxWidth().padding(vertical = 4.dp), elevation = CardDefaults.cardElevation(2.dp)) {
        Column(Modifier.padding(12.dp)) {
            Text("${vet.name} - ${vet.specialty}", style = MaterialTheme.typography.titleMedium)
            Text("Fee: $${"%.2f".format(vet.consultationFee)}")
            // Display first few available slots as example
            vet.availability.entries.firstOrNull()?.let { (date, slots) ->
                if (slots.isNotEmpty()) Text("Next Available: $date at ${slots.first()}", style = MaterialTheme.typography.bodySmall)
                else Text("No immediate slots shown", style = MaterialTheme.typography.bodySmall)
            } ?: Text("Availability not shown", style = MaterialTheme.typography.bodySmall)
            Button(onClick = { onBookClick(vet) }, modifier = Modifier.align(Alignment.End).padding(top=4.dp)) { Text("Book Appointment") }
        }
    }
}

@Composable
fun AppointmentCard(appt: Appointment) {
    Card(Modifier.fillMaxWidth().padding(vertical = 4.dp), elevation = CardDefaults.cardElevation(1.dp)) {
        Column(Modifier.padding(12.dp)) {
            Text("Patient: ${appt.patientName} with ${appt.vetName}", style = MaterialTheme.typography.titleMedium)
            Text("On: ${appt.getFormattedDateTime()} for ${appt.reason}")
            Text("Status: ${appt.status}", color = if (appt.status == "booked") Color.Blue else Color.DarkGray)
        }
    }
}

@Composable
fun PatientQueueCard(entry: PatientQueueEntry, vets: List<VetProfile>, viewModel: VetConsultationViewModel) {
    var showAssignVetDialog by remember { mutableStateOf(false) }

    if (showAssignVetDialog) {
        AssignVetDialog(entry, vets, viewModel, onDismiss = { showAssignVetDialog = false })
    }

    Card(Modifier.fillMaxWidth().padding(vertical = 4.dp), elevation = CardDefaults.cardElevation(1.dp)) {
        Column(Modifier.padding(12.dp)) {
            Text("Patient: ${entry.patientName} (User: ${entry.userId})", style = MaterialTheme.typography.titleMedium)
            Text("Arrived: ${entry.getFormattedArrivalTime()}, Reason: ${entry.reason}")
            Text("Status: ${entry.status}", fontWeight = androidx.compose.ui.text.font.FontWeight.Bold)
            entry.assignedVetName?.let { Text("Assigned to: $it") }
            if (entry.assignedVetId == null) {
                Button(onClick = { showAssignVetDialog = true }, modifier = Modifier.align(Alignment.End).padding(top=4.dp)) {
                    Text("Assign Vet")
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BookingDialog(vet: VetProfile, viewModel: VetConsultationViewModel, onDismiss: () -> Unit) {
    var selectedDate by remember { mutableStateOf(vet.availability.keys.firstOrNull() ?: "") }
    var selectedTime by remember { mutableStateOf(vet.availability[selectedDate]?.firstOrNull() ?: "") }
    var patientName by remember { mutableStateOf(viewModel.mockPatientNames.first()) }
    var reason by remember { mutableStateOf(viewModel.mockReasons.first()) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Book with ${vet.name}") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                ExposedDropdownMenuForOptions("Date", vet.availability.keys.toList(), selectedDate) { date ->
                    selectedDate = date
                    selectedTime = vet.availability[date]?.firstOrNull() ?: ""
                }
                if (selectedDate.isNotEmpty()) {
                    ExposedDropdownMenuForOptions("Time", vet.availability[selectedDate] ?: emptyList(), selectedTime) { selectedTime = it }
                }
                ExposedDropdownMenuForOptions("Patient", viewModel.mockPatientNames, patientName) { patientName = it }
                ExposedDropdownMenuForOptions("Reason", viewModel.mockReasons, reason) { reason = it }
            }
        },
        confirmButton = {
            Button(onClick = {
                if (selectedDate.isNotBlank() && selectedTime.isNotBlank()) {
                    viewModel.bookAppointment(vet.vetId, "user_currentUser", patientName, selectedDate, selectedTime, reason)
                }
                onDismiss()
            }, enabled = selectedDate.isNotBlank() && selectedTime.isNotBlank()) { Text("Book") }
        },
        dismissButton = { Button(onClick = onDismiss) { Text("Cancel") } }
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddToQueueDialog(viewModel: VetConsultationViewModel, onDismiss: () -> Unit) {
    var patientName by remember { mutableStateOf(viewModel.mockPatientNames.first()) }
    var reason by remember { mutableStateOf(viewModel.mockReasons.first()) }
    var userId by remember { mutableStateOf("farmer_walkin_user") }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Add Patient to Queue")},
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                 ExposedDropdownMenuForOptions("Patient", viewModel.mockPatientNames, patientName) { patientName = it }
                 ExposedDropdownMenuForOptions("Reason", viewModel.mockReasons, reason) { reason = it }
                 OutlinedTextField(value = userId, onValueChange = {userId = it}, label = {Text("User ID (Optional)")})
            }
        },
        confirmButton = { Button(onClick = {
            viewModel.addToQueue(userId.ifBlank { "anonymous_user" }, patientName, reason)
            onDismiss()
        }) {Text("Add to Queue")}},
        dismissButton = { Button(onClick = onDismiss) {Text("Cancel")}}
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AssignVetDialog(entry: PatientQueueEntry, vets: List<VetProfile>, viewModel: VetConsultationViewModel, onDismiss: () -> Unit){
    var selectedVetId by remember { mutableStateOf(vets.firstOrNull()?.vetId ?: "")}
    AlertDialog(
        onDismissRequest = onDismiss,
        title = {Text("Assign Vet for ${entry.patientName}")},
        text = {
            ExposedDropdownMenuForOptions("Select Vet", vets.map { "${it.name} (${it.vetId})" }, vets.find{it.vetId == selectedVetId}?.let{"${it.name} (${it.vetId})"} ?: "Select Vet") {
                selectedVetId = it.substringAfterLast("(").substringBeforeLast(")") // Extract ID
            }
        },
        confirmButton = { Button(onClick = {
            if(selectedVetId.isNotBlank()) viewModel.assignQueuedPatient(entry.queueId, selectedVetId)
            onDismiss()
        }, enabled = selectedVetId.isNotBlank()) {Text("Assign")}},
        dismissButton = { Button(onClick = onDismiss) {Text("Cancel")}}
    )
}


// Re-using ExposedDropdownMenuForOptions, assuming it's accessible
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ExposedDropdownMenuForOptions( // Copied for standalone preview if needed
    label: String,
    options: List<String>,
    selectedOption: String,
    onOptionSelected: (String) -> Unit
) {
    var expanded by remember { mutableStateOf(false) }
    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = !expanded }
    ) {
        OutlinedTextField(
            value = selectedOption,
            onValueChange = {}, readOnly = true,
            label = { Text(label) },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            modifier = Modifier.menuAnchor().fillMaxWidth()
        )
        ExposedDropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false }
        ) {
            options.forEach { option ->
                DropdownMenuItem(
                    text = { Text(option) },
                    onClick = {
                        onOptionSelected(option)
                        expanded = false
                    }
                )
            }
        }
    }
}

@Preview(showBackground = true, heightDp = 1000)
@Composable
fun PreviewVetConsultationScreen() {
    MaterialTheme {
        VetConsultationScreen(viewModel = VetConsultationViewModel())
    }
}

File: app/veterinary/vet_consultation_screen.py
import datetime
import random

class VetConsultationScreen:
    def __init__(self):
        self.vets = {} # vet_id: {name, specialty, availability}
        self.appointments = [] # List of booked appointments
        self.patient_queue = [] # Patients waiting for immediate/walk-in consultation
        self._initialize_mock_data()

    def _initialize_mock_data(self):
        # Mock Veterinarians
        vet_names = ["Dr. Alice Smith", "Dr. Bob Johnson", "Dr. Carol White", "Dr. David Brown"]
        specialties = ["General Practice", "Surgery", "Dermatology", "Livestock Specialist"]
        for i, name in enumerate(vet_names):
            vet_id = f"vet{str(i+1).zfill(3)}"
            self.vets[vet_id] = {
                "name": name,
                "specialty": random.choice(specialties),
                "availability": self._generate_mock_availability(), # {date_str: [time_slots]}
                "consultation_fee": round(random.uniform(50, 150), 2)
            }

        # Mock some existing appointments
        farmer_user_ids = [f"farmer{str(i).zfill(3)}" for i in range(1, 10)]
        pet_names = ["Bessie", "Charlie", "Daisy", "Rocky", "Lucy", "Max"]
        for i in range(5): # Create 5 mock appointments
            vet_id = random.choice(list(self.vets.keys()))
            vet_avail_dates = list(self.vets[vet_id]["availability"].keys())
            if not vet_avail_dates: continue

            appt_date_str = random.choice(vet_avail_dates)
            time_slots = self.vets[vet_id]["availability"][appt_date_str]
            if not time_slots: continue

            appt_time_str = random.choice(time_slots)
            # Attempt to book this slot (simplified)
            if self._book_slot_internal(vet_id, appt_date_str, appt_time_str):
                self.appointments.append({
                    "appointment_id": f"appt_{datetime.datetime.now().strftime('%Y%m%d%H%M')}_{str(i).zfill(3)}",
                    "vet_id": vet_id,
                    "vet_name": self.vets[vet_id]["name"],
                    "user_id": random.choice(farmer_user_ids),
                    "patient_name": random.choice(pet_names),
                    "appointment_datetime": datetime.datetime.strptime(f"{appt_date_str} {appt_time_str}", "%Y-%m-%d %H:%M"),
                    "reason": random.choice(["General check-up", "Vaccination", "Skin issue", "Limping"]),
                    "status": "booked" # Other statuses: completed, cancelled, in_progress
                })
        self.appointments.sort(key=lambda x: x["appointment_datetime"])

        # Mock patient queue (for walk-ins or immediate online requests)
        for i in range(3):
            self.patient_queue.append({
                "queue_id": f"q_{str(i).zfill(3)}",
                "user_id": random.choice(farmer_user_ids),
                "patient_name": random.choice(pet_names),
                "arrival_time": datetime.datetime.now() - datetime.timedelta(minutes=random.randint(5,30)),
                "reason": "Urgent concern",
                "assigned_vet_id": None # Vet can pick from queue
            })
        self.patient_queue.sort(key=lambda x: x["arrival_time"])


    def _generate_mock_availability(self, num_days=7, slots_per_day=5):
        availability = {}
        today = datetime.date.today()
        for i in range(num_days):
            day = today + datetime.timedelta(days=i)
            if day.weekday() >= 5: continue # Skip weekends for this mock

            time_slots = []
            for j in range(slots_per_day):
                hour = 9 + j # 9 AM to 1 PM for example
                if random.random() > 0.3: # 70% chance slot is available
                    time_slots.append(f"{hour:02d}:00")
            if time_slots:
                availability[day.isoformat()] = time_slots
        return availability

    def _book_slot_internal(self, vet_id, date_str, time_str):
        """ Helper to remove a slot. True if successful."""
        if date_str in self.vets[vet_id]["availability"] and \
           time_str in self.vets[vet_id]["availability"][date_str]:
            self.vets[vet_id]["availability"][date_str].remove(time_str)
            return True
        return False

    def get_vet_availability(self, vet_id, date_str=None):
        vet = self.vets.get(vet_id)
        if not vet: return None
        if date_str:
            return {date_str: vet["availability"].get(date_str, [])}
        return vet["availability"]

    def book_appointment(self, vet_id, user_id, patient_name, date_str, time_str, reason):
        if vet_id not in self.vets:
            print(f"Error: Vet ID {vet_id} not found.")
            return None

        available_slots_on_date = self.vets[vet_id]["availability"].get(date_str, [])
        if time_str not in available_slots_on_date:
            print(f"Error: Slot {time_str} on {date_str} for Dr. {self.vets[vet_id]['name']} is not available or date is invalid.")
            return None

        if self._book_slot_internal(vet_id, date_str, time_str):
            new_appointment = {
                "appointment_id": f"appt_{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}_{str(len(self.appointments)).zfill(3)}",
                "vet_id": vet_id,
                "vet_name": self.vets[vet_id]["name"],
                "user_id": user_id,
                "patient_name": patient_name,
                "appointment_datetime": datetime.datetime.strptime(f"{date_str} {time_str}", "%Y-%m-%d %H:%M"),
                "reason": reason,
                "status": "booked"
            }
            self.appointments.append(new_appointment)
            self.appointments.sort(key=lambda x: x["appointment_datetime"])
            print(f"Appointment booked for {patient_name} with {self.vets[vet_id]['name']} on {date_str} at {time_str}.")
            return new_appointment
        else:
            # This case should ideally be caught by the check above, but as a fallback
            print(f"Error: Could not book slot {time_str} on {date_str} for Dr. {self.vets[vet_id]['name']}. Slot may have just been taken.")
            return None

    def get_upcoming_appointments(self, user_id=None, vet_id=None, limit=10):
        results = self.appointments
        if user_id:
            results = [appt for appt in results if appt["user_id"] == user_id and appt["status"] == "booked"]
        if vet_id:
            results = [appt for appt in results if appt["vet_id"] == vet_id and appt["status"] == "booked"]
        return sorted([appt for appt in results if appt["appointment_datetime"] >= datetime.datetime.now()], key=lambda x: x["appointment_datetime"])[:limit]

    def add_to_patient_queue(self, user_id, patient_name, reason):
        queue_entry = {
            "queue_id": f"q_{datetime.datetime.now().strftime('%H%M%S')}_{str(len(self.patient_queue)).zfill(3)}",
            "user_id": user_id,
            "patient_name": patient_name,
            "arrival_time": datetime.datetime.now(),
            "reason": reason,
            "assigned_vet_id": None
        }
        self.patient_queue.append(queue_entry)
        self.patient_queue.sort(key=lambda x: x["arrival_time"])
        print(f"{patient_name} added to the patient queue.")
        return queue_entry

    def assign_queued_patient_to_vet(self, queue_id, vet_id):
        for entry in self.patient_queue:
            if entry["queue_id"] == queue_id and entry["assigned_vet_id"] is None:
                if vet_id in self.vets:
                    entry["assigned_vet_id"] = vet_id
                    entry["status"] = "assigned" # Or "waiting_for_vet"
                    print(f"Patient from queue ({queue_id}) assigned to Dr. {self.vets[vet_id]['name']}.")
                    # Potentially move to a different list or vet's personal queue here
                    return True
        print(f"Could not assign patient from queue {queue_id} to vet {vet_id}. Entry not found or already assigned.")
        return False

    def display_screen(self):
        print("---- Vet Consultation Screen ----")

        print("\n-- Veterinarians Available (Sample: Dr. Alice Smith) --")
        sample_vet_id = "vet001" # Assuming this vet exists from mock data
        if sample_vet_id in self.vets:
            print(f"  Dr. {self.vets[sample_vet_id]['name']}, Specialty: {self.vets[sample_vet_id]['specialty']}")
            availability = self.get_vet_availability(sample_vet_id)
            if availability:
                for date_str, slots in list(availability.items())[:2]: # Show for 2 days
                    if slots:
                        print(f"    Date: {date_str}, Available Slots: {', '.join(slots[:3])}{'...' if len(slots)>3 else ''}")
            else:
                print(f"    No availability shown for {self.vets[sample_vet_id]['name']}.")
        else:
            # Fallback if vet001 is not generated, pick first available vet
            if self.vets:
                first_vet_id = list(self.vets.keys())[0]
                print(f"  Showing sample for Dr. {self.vets[first_vet_id]['name']}, Specialty: {self.vets[first_vet_id]['specialty']}")
                availability = self.get_vet_availability(first_vet_id)
                if availability:
                    for date_str, slots in list(availability.items())[:2]:
                         if slots:
                            print(f"    Date: {date_str}, Available Slots: {', '.join(slots[:3])}{'...' if len(slots)>3 else ''}")
            else:
                print("  No veterinarians loaded.")


        print("\n-- Book Appointment (Example) --")
        # Find a vet with some availability to book for demo
        bookable_vet_id = None
        bookable_date = None
        bookable_time = None
        for vid, vdata in self.vets.items():
            for d, ts in vdata["availability"].items():
                if ts:
                    bookable_vet_id = vid
                    bookable_date = d
                    bookable_time = ts[0]
                    break
            if bookable_vet_id: break

        if bookable_vet_id and bookable_date and bookable_time:
            self.book_appointment(bookable_vet_id, "farmer010", "Speckles", bookable_date, bookable_time, "Routine check")
        else:
            print("  Could not find an available slot for booking demonstration.")


        print("\n-- Upcoming Appointments (Next 3) --")
        upcoming = self.get_upcoming_appointments(limit=3)
        if not upcoming:
            print("  No upcoming appointments.")
        else:
            for appt in upcoming:
                print(f"  ID: {appt['appointment_id']}, Patient: {appt['patient_name']}, Vet: {appt['vet_name']}, "
                      f"Time: {appt['appointment_datetime'].strftime('%Y-%m-%d %H:%M')}, Reason: {appt['reason']}")

        print("\n-- Patient Queue (First 3) --")
        if not self.patient_queue:
            print("  Patient queue is empty.")
        else:
            for entry in self.patient_queue[:3]:
                assign_status = f"Assigned to: {self.vets[entry['assigned_vet_id']]['name']}" if entry['assigned_vet_id'] else "Unassigned"
                print(f"  Queue ID: {entry['queue_id']}, Patient: {entry['patient_name']}, Arrived: {entry['arrival_time'].strftime('%H:%M:%S')}, "
                      f"Reason: {entry['reason']}, Status: {assign_status}")

        print("\n-- Assign Patient from Queue (Example) --")
        if self.patient_queue and self.patient_queue[0]['assigned_vet_id'] is None:
            q_id_to_assign = self.patient_queue[0]['queue_id']
            # Find a vet to assign to
            assign_to_vet_id = list(self.vets.keys())[0] if self.vets else None
            if assign_to_vet_id:
                self.assign_queued_patient_to_vet(q_id_to_assign, assign_to_vet_id)
            else:
                print("  No vets available to assign patient from queue.")
        else:
            print("  No unassigned patients in queue for assignment demonstration or queue is empty.")

        print("-----------------------------")

if __name__ == '__main__':
    screen = VetConsultationScreen()
    screen.display_screen()
